### PR TITLE
Refactor customer

### PR DESCRIPTION
**Pull request by capsert**

### CODE REVIEW

File: Mollie/API/Object/Customer.py [hunk 1]
```diff
@@ -2,4 +2,51 @@
 
 
 class Customer(Base):
-    pass
+    @property
+    def id(self):
+        if 'id' not in self:
```
Comment (whyscream) [inline Mollie/API/Object/Customer.py]: We created `self.getProperty('id')` just for this, remember? Also for all the other props

File: examples/8-list-customers.py [hunk 2]
```diff
@@ -34,12 +32,12 @@ def main():
         #
         customers = mollie.customers.all(offset=0, count=amount_of_customers_to_retrieve)
 
-        body = '<p>Your API key has %u customers.</p>' % int(customers['totalCount'])
+        body = '<p>Your API key has %u customers.</p>' % int(customers.count)
 
-        if int(customers['totalCount']) == 0:
+        if int(customers['count']) == 0:
```
Comment (whyscream) [inline examples/8-list-customers.py]: `customers.count`, just like above

File: examples/9-create-customer-payment.py [hunk 3]
```diff
@@ -4,17 +4,20 @@
 #
 from __future__ import print_function
 
-import sys, os, time, flask
+import os
+import sys
+import time
+
 from app import database_write
+import Mollie
```
Comment (whyscream) [inline examples/9-create-customer-payment.py]: `import ...` comes before `from ... import ...`

File: examples/9-create-customer-payment.py [hunk 4]
```diff
@@ -24,11 +27,11 @@ def main():
         # See: https://www.mollie.com/dashboard/settings/profiles
         #
         mollie = Mollie.API.Client()
-        mollie.setApiKey('test_bt7vvByF6jTcBR4dLuW66eNnHYNIJp')
+        mollie.setApiKey('test_sBJ8CFv4229KpNVjPA8PwMBmQxyNsU')
```
Comment (whyscream) [inline examples/9-create-customer-payment.py]: apikey

File: tests/responses/create_customer.json [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+        "self": {
+            "href": "https://api.mollie.com/v2/customers/cst_8wmqcHMN4U",
+            "type": "application/hal+json"
+        },
+        "documentation": {
+            "href": "https://docs.mollie.com/reference/v2/customers-api/create-customer",
+            "type": "text/html"
+        }
+    }
+}
```
Comment (whyscream) [inline tests/responses/create_customer.json]: missing newline

File: tests/test_customer.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
+def test_create_customers(client, response):
+    response.post('https://api.mollie.com/v2/customers', 'create_customer')
+    customer = client.customers.create({
+        'name': 'Customer A',
+        'email': 'customer@example.org',
+        'locale': 'nl_NL'
```
Comment (whyscream) [inline tests/test_customer.py]: missing trailing comma

File: examples/8-list-customers.py [hunk 7]
```diff
@@ -34,12 +32,12 @@ def main():
...
 
-        body = '<p>Your API key has %u customers.</p>' % int(customers['totalCount'])
+        body = '<p>Your API key has %u customers.</p>' % int(customers.count)
 
-        if int(customers['totalCount']) == 0:
+        if int(customers.count) == 0:
             return body
 
-        if int(customers['totalCount']) > amount_of_customers_to_retrieve:
+        if int(customers['count']) > amount_of_customers_to_retrieve:
```
Comment (whyscream) [inline examples/8-list-customers.py]: 3x is ijs halen!

**Review by whyscream — COMMENTED** [review]
I'm missing a lot of tests

File: tests/responses/all_customers.json [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+                "locale": "nl_NL",
+                "metadata": null,
+                "recentlyUsedMethods": [
+                    "creditcard",
+                    "ideal"
+                ],
+                "createdAt": "2018-04-06T13:23:21.0Z",
+                "_links": {
+                    "self": {
+                        "href": "https://api.mollie.com/v2/customers/cst_kEn1PlbGa",
```
Comment (whyscream) [inline tests/responses/all_customers.json]: customer id in href does not match object id.

File: tests/responses/all_customers.json [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        ]
+    },
+    "_links": {
+        "self": {
+            "href": "https://api.mollie.com/v2/customers",
+            "type": "application/hal+json"
+        },
+        "previous": null,
+        "next": {
+            "href": "https://api.mollie.com/v2/customers?from=cst_stTC2WHAuS",
```
Comment (whyscream) [inline tests/responses/all_customers.json]: does this match the behaviour of `count`? does `count` reflect all results, or only the ones in this paginated result set?

Comment (capsert) [reply tests/responses/all_customers.json]: Count reflects the amount of results requested

File: examples/8-list-customers.py [hunk 10]
```diff
@@ -1,19 +1,20 @@
 # coding=utf-8
 #
-#  Example 8 - Retrieving all of your customers with offset and count
+# Example 7 - How to create a new customer with the Mollie API.
```
Comment (whyscream) [inline examples/8-list-customers.py]: Wrong text?

File: tests/conftest.py [hunk 11]
```diff
@@ -25,11 +25,17 @@ def get(self, url, filename, status=200):
...
 
     def post(self, url, filename, status=200):
-        """Setup a mock response for a GET request."""
+        """Setup a mock response for a POST request."""
         file = os.path.join(os.path.dirname(__file__), 'responses', '%s.json' % filename)
         body = open(file).read()
         self.add(responses.POST, url, body=body, status=status)
 
+    def delete(self, url, filename, status=204):
+        """Setup a mock response for a UPDATE request."""
```
Comment (whyscream) [inline tests/conftest.py]: Update or delete?

File: tests/responses/update_customer.json [hunk 12]
```diff
@@ -0,0 +1,24 @@
+{
```
Comment (whyscream) [inline tests/responses/update_customer.json]: create_customer and update_customer have the exact same content and response (namely a single full customer record) Don't use 2 response mock files for that..

File: tests/test_customer.py [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+    response.delete('https://api.mollie.com/v2/customers/cst_8wmqcHMN4U', 'empty')
+    deleted_customer = client.customers.delete('cst_8wmqcHMN4U')
+    assert deleted_customer == {}
+
+
+def test_customers_all(client, response):
+    response.get('https://api.mollie.com/v2/customers', 'all_customers')
+    customers = client.customers.all()
+    assert customers.count == 3
+    itterated = 0
```
Comment (whyscream) [inline tests/test_customer.py]: grammar error

File: tests/responses/update_customer.json [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+    "resource": "customer",
+    "id": "cst_8wmqcHMN4U",
+    "mode": "test",
+    "name": "Updated Customer A",
+    "email": "updated-customer@example.org",
+    "locale": "nl_NL",
+    "metadata": null,
+    "recentlyUsedMethods": [
```
Comment (whyscream) [inline tests/responses/update_customer.json]: remove this

**Pull request merged**